### PR TITLE
ARROW-4595: [Rust] Implement Table API (a.k.a DataFrame)

### PR DESCRIPTION
**Pull request by andygrove**
This PR adds a `DataFrame` trait styled loosely after Apache Spark. I would love to get feedback on this idea, especially from the point of view of people with experience of Pandas and other DataFrame implementations.

### CODE REVIEW

**Comment by wesm** [comment]
You might look at what we've done in https://github.com/ibis-project/ibis. The Spark DataFrame API falls short of fully capturing SQL semantics and so I personally would not model a deferred / expr-based dataframe-like library after it.

**Comment by wesm** [comment]
cc @kszucs @cpcloud

**Comment by paddyhoran** [comment]
cc me

**Comment by nevi-me** [comment]
Hi @andygrove, 

I've recently been playing around with a dataframe library at https://github.com/nevi-me/rust-dataframe. I started off with eager evaluation, and I had started researching how to transition to lazy evaluation to translate to logical plans. I've outlined why I started working on it, but one can think of it as a short-term experiment to see how a Rust user would use Arrow.

I think our API might depend on whether we want a full-featured implementation (DataFusion being the primary user), or whether the API is mainly an alternative way to interact with the query backend, in which case we should/could keep it simple with just transforms.
For example, adding IO (write_csv,parquet) on the trait creates an expectation that whoever implements the trait should support those data formats.

So if the DataFrame trait is meant for DataFusion use only, then that's fine, otherwise I'd think IO methods should not live in the trait.

I think in general and the medium term; we might need to hash out what parts of DataFusion we can move into Arrow (might only be compute and scalar values?), and discuss what in your original roadmap will live where (e.g clustering, Kubernetes support).

**Comment by kszucs** [comment]
I'd prefer the adaptation of abstractions from ibis, especially because DataFusion and ibis are pretty similar. 
In my mental representation `DataFrame` is an eager evaluated container (mostly because of pandas), whereas ibis calls it `Table` which IMO better aligns with the traditional SQL systems' multi-layer, lazy execution model.

**Comment by andygrove** [comment]
Thanks for the feedback so far. I’ve been looking at Ibis docs and it looks good. I will work on creating a trait and some sample code using this approach.

@nevi-me I agree that this DataFrame/Table API should not have anything specific to csv, parquet, etc. It should be possible to call some context/provider method to load the initial Table and then call a context/provider to write a Table out. Something like this pseudo code:

```rust
let t = executionContext.parquet.read(“test.parquet);
let t2 = t.select(...).filter(....).limit(...);
executionContext.csv.write(“test.csv);
```

As for the longer term roadmap, those are good questions but I don’t really have answers yet. For now I am focusing on getting the code to a point where we can at least run real-world queries using a single thread. The next step would be parallel execution with multiple threads. Distributed would come after that.

**Comment by kszucs** [comment]
Please let me know if I can help, not with actual code writing, but with the semantics of ibis! I'm really interested in it :)

**Comment by andygrove** [comment]
@kszucs I just force pushed ... please let me know how this is looking ... I plan to start implementing this for real in the next week or two

**Comment by andygrove** [comment]
With these latest changes I can now get a `Table` from the context and perform projection and limit and get a logical plan that I can then execute.

I think it would be good to review this current design and merge it once any feedback is addressed and then I can add new methods to the Table API as separate PRs.

**Comment by andygrove** [comment]
Hi @kszucs .. just a reminder that this is ready for review. Thanks!

**Comment by andygrove** [comment]
@sunchao @paddyhoran @kszucs This PR has been open for a couple weeks now .. I have just rebased .. any objection to merging this one?

File: rust/datafusion/src/execution/context.rs [hunk 1]
```diff
@@ -100,12 +102,30 @@ impl ExecutionContext {
...
             .borrow_mut()
             .insert(name.to_string(), provider);
     }
 
+    /// Get a table by name
+    pub fn table(&mut self, table_name: &str) -> Result<Arc<Table>> {
+        match self.datasources.borrow().get(table_name) {
+            Some(provider) => {
+                Ok(Arc::new(TableImpl::new(Arc::new(LogicalPlan::TableScan {
+                    schema_name: "".to_string(),
```
Comment (paddyhoran) [inline rust/datafusion/src/execution/context.rs]: It seems odd to always use an empty string, should `schema_name` be an `Option<String>`

File: rust/datafusion/src/execution/context.rs [hunk 2]
```diff
@@ -100,12 +102,30 @@ impl ExecutionContext {
...
+                    table_name: table_name.to_string(),
+                    schema: provider.schema().clone(),
+                    projection: None,
+                }))))
+            }
+            _ => Err(ExecutionError::General(format!(
```
Comment (paddyhoran) [inline rust/datafusion/src/execution/context.rs]: Is there a test for this branch?

File: rust/datafusion/src/execution/table_impl.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        let mut expr: Vec<Expr> = Vec::with_capacity(columns.len());
+
+        for column in columns {
+            match schema.column_with_name(column) {
+                Some((i, _)) => {
+                    projection_index.push(i);
+                    expr.push(Expr::Column(i));
+                }
+                _ => {
+                    return Err(ExecutionError::InvalidColumn(format!(
```
Comment (paddyhoran) [inline rust/datafusion/src/execution/table_impl.rs]: No test for this also.

File: rust/datafusion/src/execution/table_impl.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+/// Create a new schema by applying a projection to this schema's fields
+fn projection(schema: &Schema, projection: &Vec<usize>) -> Result<Arc<Schema>> {
+    let mut fields: Vec<Field> = Vec::with_capacity(projection.len());
+    for i in projection {
+        if *i < schema.fields().len() {
+            fields.push(schema.field(*i).clone());
+        } else {
+            return Err(ExecutionError::InvalidColumn(format!(
+                "Invalid column index {} in projection",
```
Comment (paddyhoran) [inline rust/datafusion/src/execution/table_impl.rs]: No test again.

**Review by paddyhoran — APPROVED** [review]
I did not formally review this before as I wanted to defer to @kszucs due to his experience building ibis (i.e. I can speak to the design).  

To me this looks good to merge, I noted a few nits but nothing major.  In general, the tests seem more like examples than focused unit tests but we could open new JIRA's to follow up on the testing aspects.

**Comment by andygrove** [comment]
Thanks @paddyhoran I will create a follow up PR with more tests.

**Pull request closed without merging**